### DISCUSSION TITLE
pass options to setup function package

### DISCUSSION THREAD

**Discussion by yochem** [wbthomason/packer.nvim#493] [Ideas]
Hi!

I've noticed that (almost all) packages written in lua can receive configuration via a `setup()` function. For example:

```lua
require('onedark').setup({
    transparent = true,
    colors = {fg = "#ffffff", fg_dark = "#ffffff"}
})
```

My current setup is to first have packer's startup function in my init.lua and after that all `package.setup()`'s. This is irritating when packages are deleted from the packer startup function. The configs remain, and nvim will start with errors. To fix this, these setup calls can be included in packer via the `config` key:

```lua
...
use {'ful1e5/onedark.nvim', config = function() require('onedark').setup({...}) end}
...
```

But doing this for every package gets quite explicit. It would be cool if packer had a keyword for this:

```lua
...
use {'ful1e5/onedark.nvim',  setup_params = {...}}
...
```

And then internally it would run `require(package).setup(setup_params)`. What do you think of this idea? I feel like many people can take advantage of having something like this included in packer.

I hope I have explained my idea well enough, if questions remain, please ask them :)

**Comment by wbthomason** [comment]
Thanks, this is an interesting idea. I'm a little bit hesitant to add this to `packer` as - though many do - not all Lua plugins use the `setup` function pattern, and there's not much overhead to using the existing `config` key mechanism (which is more general). There's also a bit of additional complication in figuring out which Lua module to require (module names often do not match plugin names).

That said, if this is something the community wants, I'm not completely opposed to it - my concern is just avoiding undue increases to `packer`'s (already large) complexity. I'd like to get more buy-in before adding this - maybe post a link to this discussion on the Gitter, etc., and see what other people think?

**Reply by wbthomason** [reply]
Also noting that, regardless of buy-in, this is a post-#402 feature.

**Comment by yochem** [comment]
Thank you for your interest in the idea 😊. Your points are very valid. It would be great to add if a setup function would be the standard. I will look into posting this on Gitter!
